### ISSUE TITLE
Unet being cast to fp16 in LoRa training script despite contradicting comment

### ISSUE THREAD

**Issue by stefsietz** [huggingface/diffusers#3354]
### Describe the bug

[This comment](https://github.com/huggingface/diffusers/blob/0ffac97933d8ca1487d4ae9c374504a0979f0c5d/examples/text_to_image/train_text_to_image_lora.py#L419) states that "For mixed precision training we cast the text_encoder and vae weights to half-precision", but a few lines later, the Unet is also being cast. As I understand the comment, the layers through which the gradient flows to the LoRa layers should stay in fp32 precision, so this seems to be a bug.
```
# For mixed precision training we cast the text_encoder and vae weights to half-precision
# as these models are only used for inference, keeping weights in full precision is not required.
weight_dtype = torch.float32
if accelerator.mixed_precision == "fp16":
    weight_dtype = torch.float16
elif accelerator.mixed_precision == "bf16":
    weight_dtype = torch.bfloat16

# Move unet, vae and text_encoder to device and cast to weight_dtype
unet.to(accelerator.device, dtype=weight_dtype)
vae.to(accelerator.device, dtype=weight_dtype)
text_encoder.to(accelerator.device, dtype=weight_dtype)
```

### Reproduction

Reading the code is enough ;-)

### System Info

Last commit: f842396

**Comment by patrickvonplaten** [comment]
Hey @stefsietz,

Good catch! We need to update the comment here to something like: 

```
# For mixed precision training we cast all non-trainable weigths (vae, non-lora text_encoder and non-lora unet) to half-precision
# as these weights are only used for inference, keeping weights in full precision is not required.
weight_dtype = torch.float32
```

**Comment by patrickvonplaten** [comment]
Would you like to open a PR to fix it?

**Comment by eyalgutflaish** [comment]
The problem was not the comment, but the casting of the unet to float16 when gradients are propagating from it the the lora weights. 
As  I understand we should let the accelerate  mixed-precision module decide what layers in the unet should be in full or low precision.
